R/calculate_study_day.R

```diff
@@ -0,0 +1,142 @@
+#' is DM domain in most cases), and calculate the study day from reference date
+#' and target date. In case of unexpected conditions like reference date is not
+#' unique for each patient, or reference and input dates are not actual dates,
+#' NA will be returned for those records.
+#'
+#' @md
+#' @param sdtm_in input data.frame that contains the target date.
+#' @param dm_domain reference date.frame that contains the reference date.
+#' @param refdt reference date from `dm_domain` that will be used as reference to
+#' calculate the study day.
+#' @param tgdt target date from `sdtm_in` that will be used to calcualte the study
+#' day.
+#' @param study_day_var the new study day variable name in the output. For
+#' example, AESTDY for AE domain for CMSTDY for CM domain.
+#' @param merge_key character to represents the merging key between `sdtm_in` and
+#' `dm_domain`.
+#'
+#' @return a data.frame that takes all columns from `sdtm_in` and a new variable
+#' to represent the calculated study day.
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' ae <- data.frame(
+#'   USUBJID = c("study123-123", "study123-124", "study123-125"),
+#'   AESTDTC = c("2012-01-01", "2012-04-14", "2012-04-14")
+#' )
+#' dm <- data.frame(
+#'   USUBJID = c("study123-123", "study123-124", "study123-125"),
+#'   RFSTDTC = c("2012-02-01", "2012-04-14", NA)
+#' )
+#' ae$AESTDTC <- as.Date(ae$AESTDTC)
+#' dm$RFSTDTC <- as.Date(dm$RFSTDTC)
+#' calculate_study_day(ae, dm, "RFSTDTC", "AESTDTC", "AESTDY")
+#' }
+#'
+calculate_study_day <- function(sdtm_in,
+                                dm_domain = DM,
+                                refdt = "RFSTDTC",
+                                tgdt,
+                                study_day_var,
+                                merge_key = "USUBJID") {
+  assertthat::assert_that(is.data.frame(sdtm_in))
+  assertthat::assert_that(is.data.frame(dm_domain))
+  assertthat::assert_that(
+    utils::hasName(dm_domain, refdt),
+    msg = "dm_domain needs to have the variable of refdt."
+  )
+  assertthat::assert_that(
+    utils::hasName(sdtm_in, tgdt),
+    msg = "sdtm_in needs to have the variable of tgdt."
+  )
+  assertthat::assert_that(
+    utils::hasName(dm_domain, merge_key),
+    msg = "dm_domain needs to have the variable of merge_key."
+  )
+  assertthat::assert_that(
+    utils::hasName(sdtm_in, merge_key),
+    msg = "sdtm_in needs to have the variable of merge_key."
+  )
+  assertthat::assert_that(is.character(study_day_var))
+  # check tgdt and study_day_var matching, for example, CMSTDTC matches CMSTDY
+  if (gsub("DTC", "", tgdt) != gsub("DY", "", study_day_var)) {
+    warning(
+      "Target date and the returned study day doesn't match. ",
+      "Expecting matching date and study day, for example, CMENDTC and CMENDY"
+    )
+  }
+
+  original_variables <- names(sdtm_in)
+
+  if (!identical(sdtm_in, dm_domain)) {
+    dm_domain <- unique(dm_domain[c(merge_key, refdt)])
+
+    check_refdt_uniqueness <- dm_domain %>%
```

Comment:
```suggestion
    check_refdt_uniqueness <- dm_domain |>
```
Please use the base R pipe operator